File: sdk-node/ProctorSDK.js
const EventEmitter = require('events');
const ProctorConfig = require('./config/ProctorConfig');
const SessionManager = require('./session/SessionManager');
const ProctorWebsocketServer = require('./websocket/ProctorWebsocketServer');

/**
 * The main public entry point for the Proctoring SDK for Node.js.
 * This version is designed for the automated handshake workflow.
 *
 * @fires ProctorSDK#session
 *
 * @example
 * // server.js
 * const { ProctorSDK, ProctorConfig } = require('proctor-sdk');
 *
 * // 1. Configure and start the SDK on your backend server.
 * const config = new ProctorConfig(8080, '0.0.0.0');
 * const sdk = new ProctorSDK(config);
 *
 * // 2. Set up your session handler.
 * sdk.on('session', (session) => {
 * console.log(`[SDK] New session started: ${session.sessionId}`);
 * session.on('violation', (event) => {
 * // Handle violations...
 * });
 * });
 *
 * sdk.start();
 *
 * // 3. On your frontend, use the provided JavaScript snippet from the documentation.
 * // The snippet will automatically handle the installation prompt and session handshake
 * // with the installed client. The generateClientLaunchUrl() method is no longer used.
 */
class ProctorSDK extends EventEmitter {
    /**
     * @param {ProctorConfig} config The configuration object for the SDK.
     */
    constructor(config) {
        super();
        this.config = config;
        this.sessionManager = new SessionManager((session) => this.emit('session', session));
        this.server = new ProctorWebsocketServer(this.config, this.sessionManager);
    }

    /**
     * Starts the internal WebSocket server to listen for client connections.
     */
    start() {
        this.server.start();
    }

    /**
     * Stops the WebSocket server.
     */
    stop() {
        this.server.stop();
    }
    
    /**
     * NOTE: This method is now obsolete in the automated handshake workflow.
     * The developer should now use the provided frontend JavaScript snippet
     * which communicates with the installer directly.
     * This method is removed to enforce the correct integration pattern.
     */
    // generateClientLaunchUrl(sessionId) { ... }
}

module.exports = { ProctorSDK, ProctorConfig };

File: sdk-java/src/main/java/com/proctor/sdk/ProctorSDK.java
package com.proctor.sdk;

import com.proctor.sdk.config.ProctorConfig;
import com.proctor.sdk.session.ProctorSession;
import com.proctor.sdk.session.SessionManager;
import com.proctor.sdk.websocket.ProctorWebsocketServer;

import java.net.InetSocketAddress;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

/**
 * The final, production-ready entry point for the Proctoring SDK for Java.
 * This version allows the developer to provide installer links during configuration
 * and retrieve them to facilitate OS-aware downloads on their frontend.
 */
public class ProctorSDK {

    private final ProctorConfig config;
    private final SessionManager sessionManager;
    private final ProctorWebsocketServer server;
    private final List<Consumer<ProctorSession>> newSessionHandlers = new ArrayList<>();

    public ProctorSDK(ProctorConfig config) {
        this.config = config;
        this.sessionManager = new SessionManager(this::handleNewSession);
        InetSocketAddress address = new InetSocketAddress(config.getHost(), config.getPort());
        this.server = new ProctorWebsocketServer(address, this.sessionManager);
    }

    /**
     * Starts the internal WebSocket server.
     */
    public void start() {
        server.start();
    }

    /**
     * Stops the internal WebSocket server.
     * @throws InterruptedException if the thread is interrupted while stopping.
     */
    public void stop() throws InterruptedException {
        server.stop();
    }

    /**
     * Registers a handler to be called when a new student client connects.
     * @param handler A consumer that accepts the new ProctorSession.
     */
    public void onSession(Consumer<ProctorSession> handler) {
        newSessionHandlers.add(handler);
    }

    /**
     * (NEW) Returns the map of installer URLs provided during configuration.
     * The developer's backend will expose this data via an API endpoint for their
     * frontend to use for OS-specific download links.
     *
     * @return A map where the key is the OS ('windows', 'macos') and the value is the download URL.
     */
    public Map<String, String> getInstallerLinks() {
        return config.getInstallerUrls();
    }
    
    /**
     * The developer's frontend JavaScript is now responsible for initiating the session
     * by communicating with the installed client.
     * This SDK provides the necessary information via getInstallerLinks().
     */

    private void handleNewSession(ProctorSession session) {
        for (Consumer<ProctorSession> handler : newSessionHandlers) {
            try {
                handler.accept(session);
            } catch (RuntimeException e) {
                // Provides a more helpful error message for the developer using the SDK.
                System.err.println("ERROR: Unhandled exception in developer-provided onSession handler for session " 
                    + session.getSessionId() + ": " + e.getMessage());
            }
        }
    }
}

